src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+    /// input's weight.
+    ///
+    /// # Errors
+    /// Triggers an error when the public key does not contain as many correction words as
+    /// the length of the input (in bits).
+    pub fn eval(
+        &self,
+        key: &Key,
+        public: &PublicShare<Weight>,
+        input: &VidpfInput,
+    ) -> Result<ValueShare<Weight>, VidpfError> {
+        let mut state = EvalState::init_from_key(key);
+        let mut share = Weight::zero(&self.weight_parameter);
+
+        let n = input.len();
+        for _ in 0..n {
+            (state, share) = self.eval_next(key, public, input, &state)?;
+        }
+
+        Ok(ValueShare {
+            share,
+            proof: state.proof,
+        })
+    }
+
+    /// [`Vidpf::eval_next`] evaluates the `input` at the level indicated by the provided state
+    /// and returns a new state and a share of the input's weight at the next level.
+    pub fn eval_next(
+        &self,
+        key: &Key,
+        public: &PublicShare<Weight>,
+        input: &VidpfInput,
+        state: &EvalState,
+    ) -> Result<(EvalState, Weight), VidpfError> {
+        let level = state.level;
+
+        let CorrectionWord {
+            s_cw,
+            t_cw_l,
+            t_cw_r,
+            w_cw,
+        } = public.cw.get(level).ok_or(VidpfError::IndexLevel)?;
+
+        let seq_tilde = &self.prg(&state.seed);
+
+        let t_i = state.t_control_bit;
+        let sl = conditional_xor_seeds(&seq_tilde.0, s_cw, t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.2, s_cw, t_i);
+        let tl = seq_tilde.1 ^ (t_i & *t_cw_l);
+        let tr = seq_tilde.3 ^ (t_i & *t_cw_r);
+
+        let x_i = Choice::from(u8::from(input[level]));
+        let s_tilde_i = conditional_select_seed(x_i, &[sl, sr]);
+
+        let t_new = Choice::conditional_select(&tl, &tr, x_i);
+        let (seed_new, w_i) = self.convert(&s_tilde_i);
+
+        let zero = Weight::zero(&self.weight_parameter);
+        let mut y = w_i + Weight::conditional_select(&zero, w_cw, t_new);
+        y.conditional_negate(Choice::from(key.id));
+
+        let pi_i = &state.proof;
+        let cs_i = &public.cs.get(level).ok_or(VidpfError::IndexLevel)?;
+        let pi_tilde = Self::node_proof(input, level, &seed_new)?;
+        let h2_input = pi_tilde.conditional_xor(t_new, cs_i).xor(pi_i);
+        let pi_new = Self::node_proof_adjustment(h2_input).xor(pi_i);
+
+        let state_new = EvalState {
+            level: level + 1,
+            seed: seed_new,
+            t_control_bit: t_new,
+            proof: pi_new,
+        };
+
+        Ok((state_new, y))
+    }
+
+    /// [`Vidpf::open`] recovers the input's weight from additive shares produced
+    /// by aggregation servers.
+    ///
+    /// # Errors
+    /// Triggers an error when internal proofs are not equal.
+    pub fn open(a: ValueShare<Weight>, b: ValueShare<Weight>) -> Result<Weight, VidpfError> {
```

Comment:
This is unsafe to put in the public API since the user is only meant to "open" aggregate shares.

Also, elsewhere we treat shares as a length-2 array.

```suggestion
    /// Triggers an error when internal proofs are not equal.
    #[cfg(test)]
    pub(crate) fn open(value_shares: [ValueShare<Weight>; 2]) -> Result<Weight, VidpfError> {
```

Reply:
decided to remove this method.